File: Behavioral/visitor/visitor.py
class IndustryObject:
    def accept(self, visitor):
        pass

    def work(self):
        pass


class CarFactory(IndustryObject):
    def accept(self, visitor):
        visitor.visit(self)

    def work(self):
        return 500


class FurnitureFactory(IndustryObject):
    def accept(self, visitor):
        visitor.visit(self)

    def work(self):
        return 2000


class IndustryVisitor:
    def visit(self, factory: CarFactory):
        pass

    def visit(self, factory: FurnitureFactory):
        pass


class PrinterVisitor(IndustryVisitor):
    def visit(self, factory: CarFactory):
        print(f"Inspection for car factory: effectivness {factory.work()}")

    def visit(self, factory: FurnitureFactory):
        print(f"Inspection for furniture factory: effectivness {factory.work()}")


toyota = CarFactory()
amba = FurnitureFactory()
inspector = PrinterVisitor()

toyota.accept(inspector)
amba.accept(inspector)

File: Behavioral/visitor/visitor.rs

trait IndustryVisitor {
    fn visit_car_factory(&self, item: &CarFactory);
    fn visit_furniture_factory(&self, item: &FurnitureFactory);
}

trait Visitable {
    fn accept(&self, visitor: &dyn IndustryVisitor);
}

trait IndustryObject {
    fn work(&self) -> u32;
}

struct PrinterVisitor {

}

impl IndustryVisitor for PrinterVisitor {
    fn visit_car_factory(&self, item: &CarFactory) {
        println!("Inspection for car factory: effectivness {}", item.work());
    }

    fn visit_furniture_factory(&self, item: &FurnitureFactory) {
        println!("Inspection for furniture factory: effectivness {}", item.work());
    }
}

struct CarFactory {

}

struct FurnitureFactory {

}

impl IndustryObject for CarFactory {
    fn work(&self) -> u32 {
        500
    }
}

impl IndustryObject for FurnitureFactory {
    fn work(&self) -> u32 {
        2000
    }
}

impl Visitable for CarFactory {
    fn accept(&self, visitor: &dyn IndustryVisitor) {
        visitor.visit_car_factory(self);
    }
}

impl Visitable for FurnitureFactory {
    fn accept(&self, visitor: &dyn IndustryVisitor) {
        visitor.visit_furniture_factory(self);
    }
}

fn main() {
    let car_factory = CarFactory{};
    let furniture_factory = FurnitureFactory{};

    let inspector = PrinterVisitor{};

    car_factory.accept(&inspector);
    furniture_factory.accept(&inspector);
}